File: OOP/7_Function_Overriding.cpp
#include <iostream>
using namespace std;

// Base class Animal
class Animal {
public:
    // Virtual function for speak
    virtual void speak() {
        cout << "Animal makes a sound" << endl;
    }
};

// Derived class Dog inheriting from Animal
class Dog : public Animal {
public:
    // Overriding the speak method
    void speak() override {
        cout << "Dog barks" << endl;
    }
};

// Derived class Cat inheriting from Animal
class Cat : public Animal {
public:
    // Overriding the speak method
    void speak() override {
        cout << "Cat mews" << endl;
    }
};

int main() {
    // Creating objects
    Animal ani;
    Dog dog;
    Cat cat;

    // Calling speak methods
    ani.speak();  // Calls Animal's speak method
    dog.speak();  // Calls Dog's overridden speak method
    cat.speak();  // Calls Cat's overridden speak method

    return 0;
}

File: OOP/7_Function_Overriding.py




# Runtime Polymorphism -> Dynamic Binding | Overriding | Late Binding

# CompileTime Polymorphism-> Static Binding | Overloading | Early Binding




# function overrding in python:


class Animal:
    def speak(self):
        print("Animal makes a sound")

class Dog(Animal):
     # Overriding the speak method from the parent class
    def speak(self):
        print("Dog barks")

class Cat(Animal):
     # Overriding the speak method from the parent class
    def speak(self):
        print("Cat mews")


ani = Animal()
dog = Dog()
cat = Cat()

ani.speak()
dog.speak()
cat.speak()


